src/lib.rs

```diff
@@ -94,3 +93,188 @@ mod tests {
+    ///
+    /// ```
+    /// # use pathbuf::PathComponent;
+    /// # #[cfg(unix)]
+    /// # {
+    /// assert_eq!(
+    ///     PathComponent::with_sanitise("/etc/shadow"),
+    ///     PathComponent::new("etc_shadow").unwrap(),
+    /// );
+    /// # }
+    /// ```
+    ///
+    /// The sanitisation algorithm isn't considered stable.
+    /// Therefore the sanitised path could change in the future for the same input.
+    pub fn with_sanitise(component: &str) -> Self {
+        let sanitised_component = sanitize_filename_reader_friendly::sanitize(component);
+        Self::new(sanitised_component).unwrap_or_else(|| {
+            panic!(
+                "Expected a sanitised path of the original path '{}'",
+                component
+            )
+        })
+    }
+
+    fn is_valid(&self) -> bool {
+        use std::path::Component;
+
+        let mut components = self.path.components();
+        matches!(
+            (components.next(), components.next()),
+            (Some(Component::Normal(_)), None)
+        )
+    }
+}
+
+impl std::ops::Deref for PathComponent {
+    type Target = std::path::Path;
+
+    fn deref(&self) -> &Self::Target {
+        &self.path
+    }
+}
+
+impl AsRef<std::path::Path> for PathComponent {
+    fn as_ref(&self) -> &std::path::Path {
+        &self.path
+    }
+}
+
+/// This allows to push just a [`PathComponent`] to a [`std::path::PathBuf`].
+///
+/// ```
+/// use std::path::PathBuf;
+/// # use pathbuf::{pathbuf, PathComponent, PushPathComponent};
+/// # #[cfg(unix)]
+/// # {
+/// let mut path = PathBuf::new();
+/// path.push_component(PathComponent::new("foo").unwrap());
+/// path.push_component(PathComponent::new("bar.txt").unwrap());
+///
+/// assert_eq!(path, pathbuf!["foo", "bar.txt"])
+/// # }
+/// ```
+pub trait PushPathComponent {
+    fn push_component(&mut self, component: PathComponent);
+}
+
+impl PushPathComponent for PathBuf {
+    fn push_component(&mut self, component: PathComponent) {
+        self.push(component);
+    }
+}
+
+/// Creates a [`PathBuf`] containing the arguments without allowing path traversal.
+///
+/// ```
+/// # use std::path::PathBuf;
+/// # use pathbuf::pathbuf_safe;
+/// #
+/// # #[cfg(unix)]
+/// # {
+/// let user_input = "foo.txt";
+/// assert_eq!(pathbuf_safe!["tmp", user_input].unwrap(), PathBuf::from("tmp/foo.txt"));
+/// let user_input = "/etc/shadow";
+/// assert!(pathbuf_safe!["tmp", user_input].is_none());
+/// # }
+/// ```
+///
+/// When the first part is trusted, the `allow` keyword can be used.
+/// It allows the usage of multiple components and the root.
+///
+/// ```
+/// # use std::path::PathBuf;
+/// # use pathbuf::pathbuf_safe;
+/// #
+/// # #[cfg(unix)]
+/// # {
+/// let user_input = "foo.txt";
+/// assert_eq!(
+///     pathbuf_safe![allow "/var/tmp", user_input].unwrap(),
+///     PathBuf::from("/var/tmp/foo.txt"),
+/// );
+/// # }
+/// ```
+#[macro_export]
+macro_rules! pathbuf_safe {
```

Comment:
Nothing about the `pathbuf` macro is unsafe in the Rust language semantics sense. I'll also note that the code included here doesn't solve path traversal either, it simply makes it less likely because an attacker-controlled input is likely to be in the form of a single variable, which can then only have one path segment.

Probably better to just modify the original macro, rather than having two which are subtly-different.

Reply:
The definition of path traversal in [CWE-22](https://cwe.mitre.org/data/definitions/22.html) is: "The product uses external input to construct a pathname that is intended to identify a file or directory that is located underneath a restricted parent directory, but the product does not properly neutralize special elements within the pathname that can cause the pathname to resolve to a location that is outside of the restricted directory."
[OWASP](https://owasp.org/www-community/attacks/Path_Traversal) and [Wikipedia](https://en.wikipedia.org/wiki/Directory_traversal_attack) have similar definitions. According to these, path traversal is not possible with this implementation.

I am hesitant to modify the `pathbuf` macro as it would be a breaking change which would break every usage and in a lot of cases it isn't even possible to fix it with the macro.

As it is more about security than memory safety, maybe `pathbuf_secure` or `pathbuf_no_traverse`?